### PR TITLE
Make FinetuningStreamingDataset parameters more flexible

### PR DESCRIPTION
**Pull request by XiaohanZhangCMU**
Refactor build_text_dataloader and build_finetuning_dataloader to make them more flexible in terms of accepting extra keys used by StreamingDataset. The rest behavior should stay the same. For example, if the supplied extra keys to the dataloader are not defined in StreamingDataset, a normal TypeError is still expected to be thrown.

### CODE REVIEW

File: llmfoundry/data/finetuning/dataloader.py [hunk 1]
```diff
@@ -257,9 +262,15 @@ def build_finetuning_dataloader(
         dataset_constructor_args = {
             k: v
             for k, v in dataset_cfg.items()
-            if k in dataset_constructor_keys and
-            k not in {'streams', 'packing_ratio'}
+            if k in allowed_dataset_config_keys and k not in {
```
Comment (milocress) [inline llmfoundry/data/finetuning/dataloader.py:265]: Is there a way to do this without hardcoding the expected streaming keys? This requires that we manually keep track of any changes in the streaming repo and sync them to foundry whenever these args change on either side.

File: tests/data/test_dataloader.py [hunk 2]
```diff
@@ -686,7 +686,7 @@ def test_finetuning_dataloader_streaming(
         'dataset': {
             'max_seq_len': 2048,
             'decoder_only_format': True,
-            'allow_pad_trimming': False,
+            # 'allow_pad_trimming': False, # This should not be allowed by Streaming
```
Comment (milocress) [inline tests/data/test_dataloader.py:689]: should we delete this line?

**Review by milocress — COMMENTED** [review]
Deferring to @dakinggg , my intuition is there is some way of programmatically inspecting the streaming arguments and that this is preferable to hardcoding `streams`, `packing_ratio`, etc. as acceptable streaming arguments.

File: llmfoundry/data/finetuning/dataloader.py (hunk 1 above)
Comment (XiaohanZhangCMU) [reply llmfoundry/data/finetuning/dataloader.py:265]: ``streams`` needs to be hardcoded to get excluded. The others may be able to get programatically. But I am also inclined to keep the change minimal, in case it may affect any existing use cases unexpectedly.

I had to add these extra ones because this change [this](https://github.com/mosaicml/llm-foundry/blob/f034a8612708ca7b5c6a0523dbca9f03ffbf28e6/llmfoundry/data/finetuning/tasks.py#L655), breaks a unit test: tests/data/test_packing.py: test_auto_packing_with_streaming_dataloader.

File: llmfoundry/data/text_data.py [hunk 3]
```diff
@@ -328,9 +324,11 @@ def build_text_dataloader(
         if 'streams' in dataset_cfg else None,
     )
 
-    valid_streaming_text_dataset_parameters = inspect.signature(
-        StreamingTextDataset,
-    ).parameters
+    valid_streaming_text_dataset_parameters = set(
```
Comment (dakinggg) [inline llmfoundry/data/text_data.py:327]: this code turns this line and the next one into a noop right?

File: llmfoundry/data/finetuning/dataloader.py [hunk 4]
```diff
@@ -254,12 +259,23 @@ def build_finetuning_dataloader(
...
+        #         'seq_parallel_replication',
+        #         'pad_to_longest',
+        #     }
+        # }
         dataset_constructor_args = {
             k: v
             for k, v in dataset_cfg.items()
-            if k in dataset_constructor_keys and
-            k not in {'streams', 'packing_ratio'}
+            if k in set(dataset_constructor_keys).union(extraneous_keys)
```
Comment (dakinggg) [inline llmfoundry/data/finetuning/dataloader.py:276]: pretty sure you still need to exclude streams and packing_ratio because they are passing in explicitly

Comment (XiaohanZhangCMU) [reply llmfoundry/data/finetuning/dataloader.py]: That's true. didn't see that happen.

File: llmfoundry/data/text_data.py (hunk 3 above)
Comment (XiaohanZhangCMU) [reply llmfoundry/data/text_data.py]: 100%. logic is further cleaned up. plz tal.

File: llmfoundry/data/finetuning/dataloader.py [hunk 5]
```diff
@@ -198,11 +198,16 @@ def build_finetuning_dataloader(
...
         dataset_constructor_keys,
     ).union(_ALLOWED_DATASET_KEYS)
-    _validate_config(
+
+    extraneous_keys = _validate_config(
         **dataset_cfg,
         allowed_dataset_keys=allowed_dataset_config_keys,
     )
 
+    allowed_dataset_config_keys = allowed_dataset_config_keys.union(
```
Comment (dakinggg) [inline llmfoundry/data/finetuning/dataloader.py:207]: i don't think this gets used anywhere?

File: tests/data/test_dataloader.py (hunk 2 above)
Comment (XiaohanZhangCMU) [reply tests/data/test_dataloader.py]: removed.

File: llmfoundry/data/finetuning/dataloader.py (hunk 1 above)
Comment (XiaohanZhangCMU) [reply llmfoundry/data/finetuning/dataloader.py]: Yes, u r right~~ I don't need to change the line.

File: tests/data/test_dataloader.py [hunk 6]
```diff
@@ -686,7 +686,6 @@ def test_finetuning_dataloader_streaming(
         'dataset': {
             'max_seq_len': 2048,
             'decoder_only_format': True,
-            'allow_pad_trimming': False,
```
Comment (dakinggg) [inline tests/data/test_dataloader.py:689]: why did you need to remove this arg? this is a valid dataset config key

Comment (XiaohanZhangCMU) [reply tests/data/test_dataloader.py]: It was complaining this param, but not any more.

**Review by dakinggg — APPROVED** [review]
lgtm
